pkg/vendir/fetch/hg/sync.go

```diff
@@ -44,11 +46,37 @@ func (d Sync) Sync(dstPath string, tempArea ctlfetch.TempArea) (ctlconf.LockDire
 
 	defer os.RemoveAll(incomingTmpPath)
 
-	hg := NewHg(d.opts, d.log, d.refFetcher)
+	hg, err := NewHg(d.opts, d.log, d.refFetcher, tempArea)
+	if err != nil {
+		return hgLockConf, fmt.Errorf("Setting up hg: %w", err)
+	}
+	defer hg.Close()
+
+	if cachePath, ok := d.cache.Has("hg", hg.CacheID()); ok {
+		// Sync directly in the cache if needed
+		if !hg.CloneHasTargetRef(cachePath) {
+			if err := hg.SyncClone(cachePath); err != nil {
+				return hgLockConf, fmt.Errorf("Syncing hg cached clone: %w", err)
+			}
+		}
+		// fetch from cachedDir
+		if err := d.cache.CopyFrom("hg", hg.CacheID(), incomingTmpPath); err != nil {
+			return hgLockConf, fmt.Errorf("Extracting cached hg clone: %w", err)
+		}
+	} else {
+		// fetch in the target directory, and save it to cache
+		if err := hg.Clone(incomingTmpPath); err != nil {
+			return hgLockConf, fmt.Errorf("Cloning hg repository: %w", err)
+		}
+		if err := d.cache.Save("hg", hg.CacheID(), incomingTmpPath); err != nil {
```

Comment:
For OCI images we are making a decision to cache only when the image that vendir is retrieving is immutable (addressed by SHA). Does it make sense for us to also do the same thing here? If the user provides a ref that is not a SHA maybe we should not cache the repository.

Reply:
The goal here is to store as much history as we can, so further pulls are much quicker, not to store an exact version.

Reply:
The only issue I am concerned about here is that we are not clearing the caching, nor do we have an idea of how large our cache is currently. Due to this, if we save everything every time, it will cut down fetch times but is unbounded, which means it will grow in size forever.
Maybe we will have to rethink the caching process if this is a feature that we really want.
@Zebradil do you have any suggestion here?

Reply:
The caching topic is pretty much in line with what we're currently working on in myks (in mykso/myks#274, it is almost done and will be merged in the next few days, hopefully). The idea was to test the approach there and then to made a proposal to vendir.

And I was thinking about the cache cleanup approach just yesterday.

That being said, I'll take a look at this PR over the weekend and get back with some particular ideas.

cc @fritzduchardt

Reply:
We can do the following:
- copy the repo from the cache
- pull into the copy
- save the copy to the cache

Doing so add "only" a full copy of the repo, which is still way cheaper than a full clone. And it allows the cache system to reevaluate the entry size, solving uncontrolled growth of the cache.

I think the added safety (no unbounded growth) is worth the drawback, I will update my PR accordingly.